go/parquet/internal/encoding/levels.go

```diff
@@ -0,0 +1,284 @@
+// the buffer and return it yet, appending these encoded values. Returns the number
+// of values encoded.
+func (l *LevelEncoder) EncodeNoFlush(lvls []int16) int {
+	nencoded := 0
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		for _, level := range lvls {
+			if !l.rle.Put(uint64(level)) {
+				break
+			}
+			nencoded++
+		}
+	default:
+		for _, level := range lvls {
+			if l.bit.WriteValue(uint64(level), uint(l.bitWidth)) != nil {
+				break
+			}
+			nencoded++
+		}
+	}
+	return nencoded
+}
+
+// Flush flushes out any encoded data to the underlying writer.
+func (l *LevelEncoder) Flush() {
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		l.rleLen = l.rle.Flush()
+	default:
+		l.bit.Flush(false)
+	}
+}
+
+// Encode encodes the slice of definition or repetition levels based on
+// the currently configured encoding type and returns the number of
+// values that were encoded.
+func (l *LevelEncoder) Encode(lvls []int16) int {
+	nencoded := 0
+	if l.rle == nil && l.bit == nil {
+		panic("parquet: level encoders are not initialized")
+	}
+
+	switch l.encoding {
+	case format.Encoding_RLE:
+		for _, level := range lvls {
+			if !l.rle.Put(uint64(level)) {
+				break
+			}
+			nencoded++
+		}
+		l.rleLen = l.rle.Flush()
+	default:
+		for _, level := range lvls {
+			if l.bit.WriteValue(uint64(level), uint(l.bitWidth)) != nil {
+				break
+			}
+			nencoded++
+		}
+		l.bit.Flush(false)
+	}
+	return nencoded
+}
+
+// Len returns the number of bytes that were written as Run Length encoded
+// levels, this is only valid for run length encoding and will panic if using
+// deprecated bit packed encoding.
+func (l *LevelEncoder) Len() int {
+	if l.encoding != format.Encoding_RLE {
+		panic("parquet: level encoder, only implemented for RLE")
+	}
+	return l.rleLen
+}
+
+// LevelDecoder handles the decoding of repetition and definition levels from a
+// parquet file supporting bit packed and run length encoded values.
+type LevelDecoder struct {
+	bitWidth  int
+	remaining int
+	maxLvl    int16
+	encoding  format.Encoding
+	rle       *utils.RleDecoder
+	bit       *utils.BitReader
+}
+
+// SetData sets in the data to be decoded by subsequent calls by specifying the encoding type
+// the maximum level (which is what determines the bit width), the number of values expected
+// and the raw bytes to decode. Returns the number of bytes expected to be decoded.
+func (l *LevelDecoder) SetData(encoding parquet.Encoding, maxLvl int16, nbuffered int, data []byte) int {
+	l.maxLvl = maxLvl
+	l.encoding = format.Encoding(encoding)
+	l.remaining = nbuffered
+	l.bitWidth = bits.Len64(uint64(maxLvl))
+
+	switch encoding {
+	case parquet.Encodings.RLE:
+		if len(data) < 4 {
+			panic("parquet: received invalid levels (corrupt data page?)")
```

Comment:
i forget where we landed on this in past reviews, but unless we are checking at a higher level, I think a malformed parquet file could hit this spot (i.e. it might be better return an error).  I think the same potentially applies to the panics below in this method.

Reply:
modified to propagate error instead of panic